ddl/partition_test.go

```diff
@@ -139,3 +141,100 @@ func testTruncatePartition(t *testing.T, ctx sessionctx.Context, d ddl.DDL, dbIn
 	checkHistoryJobArgs(t, ctx, job.ID, &historyJobArgs{ver: v, tbl: tblInfo})
 	return job
 }
+
+func TestReorganizePartitionRollback(t *testing.T) {
+	// See issue: https://github.com/pingcap/tidb/issues/42448
+	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/ddl/reorgPartWriteReorgReplacedPartIDsFail", "return(true)"))
+	defer func() {
+		require.NoError(t, failpoint.Disable("github.com/pingcap/tidb/ddl/reorgPartWriteReorgReplacedPartIDsFail"))
+	}()
+
+	store, do := testkit.CreateMockStoreAndDomain(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec("CREATE TABLE `t1` (\n" +
+		"    `id` bigint(20) NOT NULL AUTO_INCREMENT,\n" +
+		"    `k` int(11) NOT NULL DEFAULT '0',\n" +
+		"    `c` char(120) NOT NULL DEFAULT '',\n" +
+		"    `pad` char(60) NOT NULL DEFAULT '',\n" +
+		"    PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */,\n" +
+		"    KEY `k_1` (`k`)\n" +
+		"  ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin\n" +
+		"  PARTITION BY RANGE (`id`)\n" +
+		"  (PARTITION `p0` VALUES LESS THAN (2000000),\n" +
+		"   PARTITION `p1` VALUES LESS THAN (4000000),\n" +
+		"   PARTITION `p2` VALUES LESS THAN (6000000),\n" +
+		"   PARTITION `p3` VALUES LESS THAN (8000000),\n" +
+		"   PARTITION `p4` VALUES LESS THAN (10000000),\n" +
+		"   PARTITION `p5` VALUES LESS THAN (MAXVALUE))")
+	tk.MustExec("insert into t1(k, c, pad) values (1, 'a', 'beijing'), (2, 'b', 'chengdu')")
+	ddlDone := make(chan error)
+	defer close(ddlDone)
+	go func() {
+		tk2 := testkit.NewTestKit(t, store)
+		tk2.MustExec("use test")
+		err := tk2.ExecToErr("alter table t1 reorganize partition p0, p1, p2, p3, p4 into( partition pnew values less than (10000000))")
+		ddlDone <- err
+	}()
+
+	showDDLJobStart := time.Now()
+	jobID := ""
+	schemaState := ""
+	for time.Since(showDDLJobStart) < time.Minute {
+		rows := tk.MustQuery("admin show ddl jobs where JOB_TYPE='alter table reorganize partition'").Rows()
+		if len(rows) == 0 {
+			time.Sleep(100 * time.Millisecond)
+			continue
+		}
+
+		require.Equal(t, 1, len(rows))
+		jobID = rows[0][0].(string)
+		schemaState = rows[0][4].(string)
+		if schemaState != "write reorganization" {
+			time.Sleep(500 * time.Millisecond)
+			continue
+		}
+		break
+	}
```

Comment:
One could also use a hook/callback for better control, like in [this example](https://github.com/pingcap/tidb/blob/d9149b0afa2a2099a27e41df152083083fbcfc04/ddl/reorg_partition_test.go#L930-L972) or [this example](https://github.com/pingcap/tidb/blob/d9149b0afa2a2099a27e41df152083083fbcfc04/ddl/db_partition_test.go#L4575-L4610).

Reply:
Good idea! I've updated the PR